sweepbatcher/greedy_batch_selection.go

```diff
@@ -0,0 +1,306 @@
+// new batch composed of this sweep only, and about all existing batches. Then
+// it passes the data to selectBatch() function, which emulates adding the sweep
+// to each batch and creating new batch for the sweep, and calculates the costs
+// of each alternative. Based on the estimates of selectBatch(), this method
+// adds the sweep to a batch, or creates new batch for it. If this method fails
+// for whatever reason, the caller falls back to the simple algorithm (method
+// handleSweep).
+func (b *Batcher) greedyAddSweep(ctx context.Context, sweep *sweep) error {
+	if b.customFeeRate == nil {
+		return errors.New("greedy batch selection algorithm requires " +
+			"setting custom fee rate provider")
+	}
+
+	// Collect weight and fee rate info about the sweep and new batch.
+	sweepFeeDetails, newBatchFeeDetails, err := estimateSweepFeeIncrement(
+		sweep,
+	)
+	if err != nil {
+		return fmt.Errorf("failed to estimate tx weight for "+
+			"sweep %x: %w", sweep.swapHash[:6], err)
+	}
+
+	// Collect weight and fee rate info about existing batches.
+	batches := make([]feeDetails, 0, len(b.batches))
+	for _, batch := range b.batches {
+		newBatchFeeDetails, err := estimateBatchWeight(batch)
+		if err != nil {
+			return fmt.Errorf("failed to estimate tx weight for "+
+				"batch %d: %w", batch.id, err)
+		}
+		batches = append(batches, newBatchFeeDetails)
+	}
+
+	// Run the algorithm. Get batchId of the best batch, or specialId if the
+	// best option is to create new batch.
+	batchId, err := selectBatch(
+		batches, sweepFeeDetails, newBatchFeeDetails,
+	)
+	if err != nil {
+		return fmt.Errorf("batch selection algorithm failed for sweep "+
+			"%x: %w", sweep.swapHash[:6], err)
+	}
+
+	// If the best option is to start new batch, do it.
+	if batchId == specialId {
+		return b.launchNewBatch(ctx, sweep)
+	}
+
+	// Locate the batch to add the sweep to.
+	batch, has := b.batches[batchId]
+	if !has {
+		return fmt.Errorf("batch selection algorithm returned "+
+			"batch id %d which doesn't exist, for sweep %x",
+			batchId, sweep.swapHash[:6])
+	}
+
+	// Add the sweep to the batch.
+	accepted, err := batch.addSweep(ctx, sweep)
+	if err != nil {
+		return fmt.Errorf("batch selection algorithm returned "+
+			"batch id %d for sweep %x, but adding failed: %w",
+			batchId, sweep.swapHash[:6], err)
+	}
+	if !accepted {
+		return fmt.Errorf("batch selection algorithm returned "+
+			"batch id %d for sweep %x, but acceptance failed",
+			batchId, sweep.swapHash[:6])
+	}
+
+	return nil
+}
+
+// estimateSweepFeeIncrement returns fee details for adding the sweep to a batch
+// and for creating new batch with this sweep only.
+func estimateSweepFeeIncrement(s *sweep) (feeDetails, feeDetails, error) {
+	// Create a fake batch with this sweep.
+	batch := &batch{
+		rbfCache: rbfCache{
+			FeeRate: s.minFeeRate,
+		},
+		sweeps: map[lntypes.Hash]sweep{
+			s.swapHash: *s,
+		},
+	}
+
+	// Estimate new batch.
+	newBatch, err := estimateBatchWeight(batch)
```

Comment:
nit: to avoid confusing names and improve readability, consider naming `newBatch` => `feeDetails1` and `batch2` => `feeDetails2` or something similar as these are not batches.

Reply:
I renamed them to `fd1` and `fd2`.